### PR TITLE
[Serve] Support scale replica down to 0

### PR DESCRIPTION
**Pull request by sihanwang41**
Add mechanism to scale replica down to 0 and scale up to 1.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
if this is ready for review, can you:
- drop the WIP label
- Clean up the code
- Assign edward and I for review

File: python/ray/serve/autoscaling_policy.py [hunk 1]
```diff
@@ -68,7 +68,10 @@ def __init__(self, config: AutoscalingConfig):
 
     @abstractmethod
     def get_decision_num_replicas(
-        self, current_num_ongoing_requests: List[float], curr_target_num_replicas: int
+        self,
+        current_num_ongoing_requests: List[float],
+        curr_target_num_replicas: int,
+        current_handle_queued_queries: float,
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_policy.py:74]: can we message the order a bit? the current ordering is (metrics, config, metrics)

File: python/ray/serve/autoscaling_policy.py [hunk 2]
```diff
@@ -119,8 +124,15 @@ def __init__(self, config: AutoscalingConfig):
...
     ) -> int:
+
+        if current_handle_queued_queries > 0 and len(current_num_ongoing_requests) == 0:
+            return max(1, curr_target_num_replicas)
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_policy.py:134]: how do we deal with the edge case of the ongoing_requests is like 0.01 due to window averages?

File: python/ray/serve/client.py [hunk 3]
```diff
@@ -120,6 +120,12 @@ def shutdown(self) -> None:
         Shuts down all processes and deletes all state associated with the
         instance.
         """
+
+        # shutdown handles
+        for k in list(self.handle_cache):
+            self.handle_cache[k].stop_metrics_pusher()
+            del self.handle_cache[k]
```
Comment (simon-mo) [inline python/ray/serve/client.py:127]: do we need explicit del here?

**Review by simon-mo — COMMENTED** [review]
can you also add more unit testing of the algorithm in `test_autoscaling_policy.py`

File: python/ray/serve/tests/test_deploy.py [hunk 4]
```diff
@@ -833,15 +833,6 @@ class BadNumReplicas:
...
 
-    with pytest.raises(ValidationError):
-
-        @serve.deployment(num_replicas=0)
-        class ZeroNumReplicas:
-            pass
-
-    with pytest.raises(ValidationError):
-        Base.options(num_replicas=0)
-
```
Comment (edoakes) [inline python/ray/serve/tests/test_deploy.py:844]: hmm `num_replicas=0` should still be invalid at the API level without autoscaling.

**Comment by edoakes** [comment]
Merge conflicts FYI

File: python/ray/serve/autoscaling_policy.py (hunk 2 above)
Comment (sihanwang41) [reply python/ray/serve/autoscaling_policy.py:134]: When ongoing_requests is bigger than 0, this if check won't take effect, the desired replicas will be calculated from `calculate_desired_num_replicas` (old behavior, this pr is not going to change the old behavior)

File: python/ray/serve/client.py (hunk 3 above)
Comment (sihanwang41) [reply python/ray/serve/client.py:127]: This is for optimization. I want the handle reference goes to 0 as soon as possible. And the background thread is going to be dead and controller is going to be shut down. I think we are good  to removed from the cache side.

File: python/ray/serve/api.py [hunk 5]
```diff
@@ -453,6 +453,9 @@ def deployment(
             "_autoscaling_config is provided."
         )
 
+    if num_replicas == 0:
+        raise ValueError("num_replicas is expected to larger than 0")
```
Comment (simon-mo) [inline python/ray/serve/api.py:457]: this conflict with the `NonNegativeInt` requirement you changed in DeploymentConfig
https://github.com/sihanwang41/ray/blob/219997e449/python/ray/serve/config.py#L111

which one is true? we shouldn't have two validation pass that disagree with each other

File: python/ray/serve/autoscaling_metrics.py [hunk 6]
```diff
@@ -1,17 +1,21 @@
...
 import bisect
 from collections import defaultdict
+import logging
 import time
 from typing import Callable, DefaultDict, Dict, List, Optional
 from dataclasses import dataclass, field
 
 import ray
 
+logger = logging.getLogger(__file__)
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_metrics.py:11]: Let's use the Serve logger, see pattern: https://github.com/sihanwang41/ray/blob/219997e449/python/ray/serve/controller.py#L26-L39

File: python/ray/serve/autoscaling_metrics.py [hunk 7]
```diff
@@ -11,7 +11,8 @@
 def start_metrics_pusher(
     interval_s: float,
     collection_callback: Callable[[], Dict[str, float]],
-    controller_handle,
+    metrics_process_func,
```
[the remark this answers is not in the document]
Comment (simon-mo) [reply python/ray/serve/autoscaling_metrics.py:17]: type annotation still missing

[the remark this answers is not in the document]
Comment (simon-mo) [reply python/ray/serve/autoscaling_metrics.py:17]: looks like if we move `.remote` to the caller side as well, the implementation can just be 
```python
def send_once():
    return metrics_process_func(data=.....)
```
and we can type it as `Callable[[Dict[str,float], float], ray.ObjectRef]]`

File: python/ray/serve/autoscaling_metrics.py [hunk 8]
```diff
@@ -1,17 +1,21 @@
...
 
+logger = logging.getLogger(__file__)
+
 
 def start_metrics_pusher(
     interval_s: float,
     collection_callback: Callable[[], Dict[str, float]],
-    controller_handle,
+    metrics_process_func,
+    stop_event=None,
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_metrics.py:18]: type annotation

File: python/ray/serve/autoscaling_metrics.py [hunk 9]
```diff
@@ -24,39 +28,51 @@ def start_metrics_pusher(
...
-
-            if last_ref:
-                ready_refs, _ = ray.wait([last_ref], timeout=0)
-                last_send_succeeded = len(ready_refs) == 1
-            if last_send_succeeded:
-                last_ref = send_once()
+            if stop_event and stop_event.is_set():
+                return
+
+            if ray.is_initialized():
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_metrics.py:50]: why is this necessary?

File: python/ray/serve/autoscaling_metrics.py [hunk 10]
```diff
@@ -24,39 +28,51 @@ def start_metrics_pusher(
...
+                        last_send_succeeded = len(ready_refs) == 1
+                    if last_send_succeeded:
+                        last_ref = send_once()
+                except Exception as e:
+                    logger.warning(
+                        "Autoscaling metrics pusher thread "
+                        "is having issue to send metrics: %s",
+                        e,
+                    )
+                    pass
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_metrics.py:63]: remove

File: python/ray/serve/autoscaling_metrics.py [hunk 11]
```diff
@@ -24,39 +28,51 @@ def start_metrics_pusher(
...
+                    if last_ref:
+                        ready_refs, _ = ray.wait([last_ref], timeout=0)
+                        last_send_succeeded = len(ready_refs) == 1
+                    if last_send_succeeded:
+                        last_ref = send_once()
+                except Exception as e:
+                    logger.warning(
+                        "Autoscaling metrics pusher thread "
+                        "is having issue to send metrics: %s",
+                        e,
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_metrics.py:61]: use f-string https://realpython.com/python-f-strings/

File: python/ray/serve/autoscaling_metrics.py [hunk 12]
```diff
@@ -24,39 +28,51 @@ def start_metrics_pusher(
...
             if remaining_time > 0:
                 time.sleep(remaining_time)
 
-    timer = threading.Thread(target=send_forever)
+    timer = threading.Thread(target=send_forever, args=[stop_event])
     # Making this a daemon thread so it doesn't leak upon shutdown, and it
     # doesn't need to block the replica's shutdown.
     timer.setDaemon(True)
     timer.start()
+    return timer
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_metrics.py:75]: now we returned a timer, please type annotation the function signature and document it.

File: python/ray/serve/autoscaling_policy.py (hunk 2 above)
Comment (simon-mo) [reply python/ray/serve/autoscaling_policy.py:134]: can you document what does this section do in code?

Comment (simon-mo) [reply python/ray/serve/autoscaling_policy.py:134]: also should this be nested with condition `len(current_num_ongoing_requests) == 0:` below for readability? so the logic goes
```
if len(current_num_ongoing_requests) == 0:
    if current_handle_queued_queries > 0: # scale up scenario
        return max(...)
    return curr_target_num_replicas
```

File: python/ray/serve/autoscaling_policy.py [hunk 13]
```diff
@@ -119,8 +124,15 @@ def __init__(self, config: AutoscalingConfig):
...
+        curr_target_num_replicas: int,
+        current_num_ongoing_requests: List[float],
+        current_handle_queued_queries: float,
     ) -> int:
+
+        if current_handle_queued_queries > 0 and len(current_num_ongoing_requests) == 0:
+            return max(1, curr_target_num_replicas)
+
         if len(current_num_ongoing_requests) == 0:
             return curr_target_num_replicas
```
Comment (simon-mo) [inline python/ray/serve/autoscaling_policy.py:137]: this does ensure proper scale down to 0 actually?

File: python/ray/serve/tests/test_autoscaling_policy.py [hunk 14]
```diff
@@ -683,6 +725,23 @@ def __call__(self):
     # Make sure start time did not change for the deployment
     assert get_deployment_start_time(controller, A) == start_time
 
+    # scale down to 0
```
Comment (simon-mo) [inline python/ray/serve/tests/test_autoscaling_policy.py:728]: also test scaling back up?

File: python/ray/serve/api.py (hunk 5 above)
Comment (sihanwang41) [reply python/ray/serve/api.py:457]: This is because we are using the DeploymentConfig for autoscaling and non autoscaling. If we are okay to receive num_replicas == 0 for non autoscaling, I can remove it.

File: python/ray/serve/autoscaling_policy.py (hunk 13 above)
Comment (sihanwang41) [reply python/ray/serve/autoscaling_policy.py:137]: scale down to 0 is ensured by following the logics in this function. When len(current_num_ongoing_requests) is 0, it means there is no replica

File: python/ray/serve/api.py (hunk 5 above)
Comment (shrekris-anyscale) [reply python/ray/serve/api.py:457]: That makes sense to me. In the future, we may want to introduce a new field or refactor the codebase, so `num_replicas` isn't so overloaded. Right now it represents both the user-configured `num_replicas` as well as the current number of replicas set by autoscaling.

File: python/ray/serve/autoscaling_metrics.py [hunk 15]
```diff
@@ -1,17 +1,21 @@
...
+
 
 def start_metrics_pusher(
     interval_s: float,
     collection_callback: Callable[[], Dict[str, float]],
-    controller_handle,
+    metrics_process_func,
+    stop_event=None,
 ):
     """Start a background thread to push metrics to controller.
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:20]: Can you add some info about what conditions cause this thread to stop, now that we've introduce a `stop_event`?

File: python/ray/serve/autoscaling_metrics.py [hunk 16]
```diff
@@ -24,39 +28,51 @@ def start_metrics_pusher(
...
+            if ray.is_initialized():
+                try:
+                    if last_ref:
+                        ready_refs, _ = ray.wait([last_ref], timeout=0)
+                        last_send_succeeded = len(ready_refs) == 1
+                    if last_send_succeeded:
+                        last_ref = send_once()
+                except Exception as e:
+                    logger.warning(
+                        "Autoscaling metrics pusher thread "
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:59]: I think we should elaborate on this error a bit. How about changing it to `"Autoscaling metrics-pusher thread failing to submit metrics to the controller, so [deployment_name] deployment cannot autoscale correctly. [Add info about what the user can do]: %s"`

If the deployment name can't be obtained here (or if it requires too much refactoring), we can omit it.

File: python/ray/serve/autoscaling_metrics.py [hunk 17]
```diff
@@ -85,6 +101,18 @@ def add_metrics_point(self, data_points: Dict[str, float], timestamp: float):
             # Using in-sort to insert while maintaining sorted ordering.
             bisect.insort(a=self.data[name], x=TimeStampedValue(timestamp, value))
 
+    def _get_datapoints(self, key: str, window_start_timestamp_s: float) -> List[float]:
+        """Get all data points given key after window_start_timestamp_s"""
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:105]:
```suggestion
        """Get all data points given key after window_start_timestamp_s"""
        
```

File: python/ray/serve/autoscaling_metrics.py [hunk 18]
```diff
@@ -102,19 +130,30 @@ def window_average(
...
+    def max(self, key: str, window_start_timestamp_s: float):
+        """Perform a max operation for metric `key`
+
+        Args:
+            key(str): the metric name.
+            window_start_timestamp_s(float): the unix epoch timestamp for the
+              start of the window. The computed average will use all datapoints
+              from this timestamp until now.
+        Returns:
+            max value of the data pointes for the key on and after time
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:151]:
```suggestion
            Max value of the data points for the key on and after time
```

File: python/ray/serve/autoscaling_metrics.py (hunk 9 above)
Comment (sihanwang41) [reply python/ray/serve/autoscaling_metrics.py:50]: If ray is not initiated (I randomly hit the issue when doing unit tests), the background thread will hit worker connect issue. 
This change is for putting some efforts to make sure ray is always initiated before sending metrics, although it is not always guaranteed.

File: python/ray/serve/autoscaling_policy.py [hunk 19]
```diff
@@ -77,6 +80,8 @@ def get_decision_num_replicas(
                 ongoing requests for each replica.
             curr_target_num_replicas (int): The number of replicas that the
                 deployment is currently trying to scale to.
+            current_handle_queued_queries (float): The number of handle queued queries,
+            if there are multiple handles, the max one should be passed in
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_policy.py:84]:
```suggestion
            if there are multiple handles, the max number of queries at a single handle should be passed in.
```

File: python/ray/serve/client.py [hunk 20]
```diff
@@ -120,6 +120,12 @@ def shutdown(self) -> None:
         Shuts down all processes and deletes all state associated with the
         instance.
         """
+
+        # shutdown handles
```
Comment (shrekris-anyscale) [inline python/ray/serve/client.py:124]:
```suggestion
        # Shut down handles
```

File: python/ray/serve/controller.py [hunk 21]
```diff
@@ -131,6 +132,9 @@ def check_alive(self) -> None:
     def record_autoscaling_metrics(self, data: Dict[str, float], send_timestamp: float):
```
Comment (shrekris-anyscale) [inline python/ray/serve/controller.py:132]: I think we should collapse `record_autoscaling_metrics` and `record_handle_metrics` into a single function and pass in the `metrics_store` as an argument.

File: python/ray/serve/controller.py [hunk 22]
```diff
@@ -201,15 +205,25 @@ def autoscale(self) -> None:
...
             new_deployment_config = deployment_config.copy()
 
             decision_num_replicas = autoscaling_policy.get_decision_num_replicas(
-                current_num_ongoing_requests=current_num_ongoing_requests,
                 curr_target_num_replicas=deployment_config.num_replicas,
+                current_num_ongoing_requests=current_num_ongoing_requests,
+                current_handle_queued_queries=current_handle_queued_queries,
             )
+
+            if decision_num_replicas == deployment_config.num_replicas:
```
Comment (shrekris-anyscale) [inline python/ray/serve/controller.py:224]: Is this for optimization?

File: python/ray/serve/deployment.py [hunk 23]
```diff
@@ -297,6 +297,16 @@ def options(
         unchanged from the existing deployment.
         """
         new_config = self._config.copy()
+
+        if num_replicas is not None and _autoscaling_config is not None:
```
Comment (shrekris-anyscale) [inline python/ray/serve/deployment.py:301]: I don't think we need these checks here. Since `options` is a property of `Deployment`, I think the checks in the `deployment` decorator definition should be sufficient.

File: python/ray/serve/deployment_state.py [hunk 24]
```diff
@@ -943,6 +943,7 @@ def __init__(
         self._curr_status_info: DeploymentStatusInfo = DeploymentStatusInfo(
             self._name, DeploymentStatus.UPDATING
         )
+        self._deployment_to_delete = False
```
Comment (shrekris-anyscale) [inline python/ray/serve/deployment_state.py:946]: `_deployment_to_delete` -> `_deleting`

File: python/ray/serve/deployment_state.py [hunk 25]
```diff
@@ -1330,12 +1331,11 @@ def _check_curr_status(self) -> bool:
...
             == 0
         ):
-            # Check for deleting.
-            if target_replica_count == 0 and all_running_replica_cnt == 0:
+            if self._deployment_to_delete:
                 return True
 
             # Check for a non-zero number of deployments.
-            elif target_replica_count == running_at_target_version_replica_cnt:
+            if target_replica_count == running_at_target_version_replica_cnt:
```
Comment (shrekris-anyscale) [inline python/ray/serve/deployment_state.py:1338]: Why change from `elif` to `if` here?

File: python/ray/serve/handle.py [hunk 26]
```diff
@@ -104,13 +105,27 @@ def __init__(
 
         self.router: Router = _router or self._make_router()
 
+        self._stop_event = threading.Event()
+        self._pusher = start_metrics_pusher(
+            interval_s=10,  # Queue handle metrcis
```
Comment (shrekris-anyscale) [inline python/ray/serve/handle.py:110]: Is this interval too large? On average, does this means the cold start delay for autoscaling increases by 5 seconds?

File: python/ray/serve/handle.py [hunk 27]
```diff
@@ -104,13 +105,27 @@ def __init__(
...
+            collection_callback=self._collect_handle_queue_metrics,
+            metrics_process_func=self.controller_handle.record_handle_metrics,
+            stop_event=self._stop_event,
+        )
+
+    def _collect_handle_queue_metrics(self):
```
Comment (shrekris-anyscale) [inline python/ray/serve/handle.py:116]:
```suggestion
    def _collect_handle_queue_metrics(self) -> Dict[str, int]:
```

File: python/ray/serve/handle.py [hunk 28]
```diff
@@ -201,6 +216,9 @@ def __reduce__(self):
     def __getattr__(self, name):
         return self.options(method_name=name)
 
+    def __del__(self):
+        self.stop_metrics_pusher()
```
Comment (shrekris-anyscale) [inline python/ray/serve/handle.py:220]: Can this throw an error? And in that case, what happens?

File: python/ray/serve/tests/test_autoscaling_policy.py [hunk 29]
```diff
@@ -178,7 +178,7 @@ def test_e2e_basic_scale_up_down(serve_instance):
     @serve.deployment(
         _autoscaling_config={
             "metrics_interval_s": 0.1,
-            "min_replicas": 1,
+            "min_replicas": 0,
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_autoscaling_policy.py:181]: Can we parametrize this test with `min_replicas=0` and `min_replicas=1`? It would help us distinguish whether future changes break only scale-to-zero or autoscaling in general.

File: python/ray/serve/tests/test_autoscaling_policy.py [hunk 30]
```diff
@@ -245,7 +245,7 @@ def test_upscale_downscale_delay():
     downscale_delay_s = 600.0
 
     config = AutoscalingConfig(
-        min_replicas=1,
+        min_replicas=0,
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_autoscaling_policy.py:248]: Same comment as [above](https://github.com/ray-project/ray/pull/24892/files#r883913049).

File: python/ray/serve/tests/test_autoscaling_policy.py [hunk 31]
```diff
@@ -259,15 +259,27 @@ def test_upscale_downscale_delay():
 
     overload_requests = [100]
 
+    # scale up when there is 0 replica and current_handle_queued_queries > 0
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_autoscaling_policy.py:262]:
```suggestion
    # Scale up when there is 0 replica and current_handle_queued_queries > 0
```

File: python/ray/serve/tests/test_autoscaling_policy.py [hunk 32]
```diff
@@ -259,15 +259,27 @@ def test_upscale_downscale_delay():
...
+    new_num_replicas = policy.get_decision_num_replicas(
+        current_num_ongoing_requests=[],
+        curr_target_num_replicas=0,
+        current_handle_queued_queries=1,
+    )
+    assert new_num_replicas == 1
+
     # We should scale up only after enough consecutive scale-up decisions.
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_autoscaling_policy.py:270]: Is this true for autoscaling in general? Do we wait for a certain number of scale up decisions?

**Review by shrekris-anyscale — CHANGES_REQUESTED** [review]
Great work so far on this change! I've left a few comments. Could we also add a unit test that checks the delay for scaling up from 0 to 1 (and scaling down from 1 to 0)? It would give us a bound on the cold start time for a simple example.

File: python/ray/serve/deployment.py (hunk 23 above)
Comment (shrekris-anyscale) [reply python/ray/serve/deployment.py:301]: Resolved asynchronously with @sihanwang41– without this check, users could overwrite their deployment's num_replicas and _autoscaling_config at the same time, which should not be allowed.

File: python/ray/serve/handle.py (hunk 28 above)
Comment (sihanwang41) [reply python/ray/serve/handle.py:220]: The stop_metrics_pusher is not supposed to throw errors. GC.should call _del_ func successfully.

File: python/ray/serve/tests/test_autoscaling_policy.py (hunk 29 above)
Comment (sihanwang41) [reply python/ray/serve/tests/test_autoscaling_policy.py:181]: good idea

File: python/ray/serve/tests/test_autoscaling_policy.py (hunk 30 above)
Comment (sihanwang41) [reply python/ray/serve/tests/test_autoscaling_policy.py:248]: this for testing multiple scenarios. it is already covered existing replicas to scale and 0 replicas to scale up

File: python/ray/serve/controller.py (hunk 21 above)
Comment (sihanwang41) [reply python/ray/serve/controller.py:132]: I will have it when we add one more store (for now the func is trivial, so I will keep it as it is). :) I am thinking to have store manager in the following pr.

File: python/ray/serve/controller.py (hunk 22 above)
Comment (sihanwang41) [reply python/ray/serve/controller.py:224]: yes, we don't need to call deploy when num_replicas is not changed.

File: python/ray/serve/deployment_state.py (hunk 25 above)
Comment (sihanwang41) [reply python/ray/serve/deployment_state.py:1338]: ha, I don't see why we need to use elif, since the first if is `return` and there is no relation between the two if now.

File: python/ray/serve/handle.py (hunk 26 above)
Comment (sihanwang41) [reply python/ray/serve/handle.py:110]: that is correct,  change it to use config number.

File: python/ray/serve/tests/test_autoscaling_policy.py (hunk 32 above)
Comment (sihanwang41) [reply python/ray/serve/tests/test_autoscaling_policy.py:271]: yes, it is true for scale down to 0.

File: python/ray/serve/autoscaling_metrics.py (hunk 9 above)
Comment (simon-mo) [reply python/ray/serve/autoscaling_metrics.py:59]: ideally this shouldn't happen (user can't get a handle before ray isn't initialized).. this is probably due to the fact that we shutdown ray between tests but the thread remains?

File: python/ray/serve/autoscaling_metrics.py (hunk 16 above)
Comment (simon-mo) [reply python/ray/serve/autoscaling_metrics.py:68]: +1

**Review by simon-mo — APPROVED** [review]
Nice work!

File: python/ray/serve/autoscaling_metrics.py [hunk 33]
```diff
@@ -1,62 +1,86 @@
...
+    metrics_process_func: Callable[[Dict[str, float], float], ray.ObjectRef],
+    stop_event: Type[Event] = None,
 ):
     """Start a background thread to push metrics to controller.
 
     We use this background so it will be not blocked by user's code and ensure
     consistently metrics delivery. Python GIL will ensure that this thread gets
     fair timeshare to execute and run.
 
+    For handle background thread usecase, when the handle is GCed or shutting down
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:30]:
```suggestion
    stop_event is passed in only when a RayServeHandle calls this function to push metrics
    for scale-to-zero. stop_event is set either when the handle is garbage collected or when
    the Serve application shuts down.
```

File: python/ray/serve/autoscaling_metrics.py [hunk 34]
```diff
@@ -1,62 +1,86 @@
...
+    the serve instance, stop_event is supposed to be set true.
+
     Args:
         interval_s(float): the push interval.
         collection_callback: a callable that returns the metric data points to
           be sent to the the controller. The collection callback should take
           no argument and returns a dictionary of str_key -> float_value.
-        controller_handle: actor handle to Serve controller.
+        metrics_process_func: actor handle function.
+        stop_event: the daemon thread will be closed when the stop event is set
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:39]:
```suggestion
        stop_event: the background thread will be closed when the stop event is set
```

File: python/ray/serve/autoscaling_metrics.py (hunk 9 above)
Comment (shrekris-anyscale) [reply python/ray/serve/autoscaling_metrics.py:59]: +1, this is probably an issue on the testing side. If possible, it would be nice to remove this if-condition since it doesn't seem necessary. If we can't, we should add a comment explaining that it's there to satisfy the unit tests.

File: python/ray/serve/autoscaling_metrics.py [hunk 35]
```diff
@@ -1,62 +1,86 @@
...
+    Returns:
+        timer: return the metrics pusher background thread object
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:41]:
```suggestion
        The background thread created by this function to push metrics to the controller.
```

File: python/ray/serve/autoscaling_metrics.py [hunk 36]
```diff
@@ -85,6 +109,19 @@ def add_metrics_point(self, data_points: Dict[str, float], timestamp: float):
...
 
+    def _get_datapoints(self, key: str, window_start_timestamp_s: float) -> List[float]:
+        """Get all data points given key after window_start_timestamp_s"""
+
+        datapoints = self.data[key]
+
+        idx = bisect.bisect(
+            a=datapoints,
+            x=TimeStampedValue(
+                timestamp=window_start_timestamp_s, value=0  # dummy value
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:120]: Can we make `TimeStampedValue`'s `value` field `Optional` instead of using a dummy value here?

File: python/ray/serve/autoscaling_metrics.py [hunk 37]
```diff
@@ -102,19 +139,30 @@ def window_average(
...
 
         if do_compact:
             self.data[key] = points_after_idx
 
         if len(points_after_idx) == 0:
             return
         return sum(point.value for point in points_after_idx) / len(points_after_idx)
+
+    def max(self, key: str, window_start_timestamp_s: float):
+        """Perform a max operation for metric `key`
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:152]:
```suggestion
        """Perform a max operation for metric `key`.
```

File: python/ray/serve/autoscaling_metrics.py [hunk 38]
```diff
@@ -102,19 +139,30 @@ def window_average(
...
+        Returns:
+            Max value of the data points for the key on and after time
+            window_start_timestamp_s, or None if there are no such points.
+        """
+        points_after_idx = self._get_datapoints(key, window_start_timestamp_s)
+
+        if len(points_after_idx) == 0:
+            return
+
+        return max(point.value for point in points_after_idx)
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:168]:
```suggestion
        return max((point.value for point in points_after_idx), default=None)
```

[`max` takes in a default](https://docs.python.org/3/library/functions.html#max) in case the iterable is empty. With this, we can remove the above lines:

```
if len(points_after_idx) == 0:
    return
```

File: python/ray/serve/autoscaling_metrics.py [hunk 39]
```diff
@@ -102,19 +139,30 @@ def window_average(
...
 
         if len(points_after_idx) == 0:
             return
         return sum(point.value for point in points_after_idx) / len(points_after_idx)
+
+    def max(self, key: str, window_start_timestamp_s: float):
+        """Perform a max operation for metric `key`
+
+        Args:
+            key(str): the metric name.
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_metrics.py:155]:
```suggestion
            key (str): the metric name.
```

File: python/ray/serve/autoscaling_policy.py [hunk 40]
```diff
@@ -119,9 +125,16 @@ def __init__(self, config: AutoscalingConfig):
...
     def get_decision_num_replicas(
-        self, current_num_ongoing_requests: List[float], curr_target_num_replicas: int
+        self,
+        curr_target_num_replicas: int,
+        current_num_ongoing_requests: List[float],
+        current_handle_queued_queries: float,
     ) -> int:
+
         if len(current_num_ongoing_requests) == 0:
+            # When 0 replica and queries queued, scale up the replicas
```
Comment (shrekris-anyscale) [inline python/ray/serve/autoscaling_policy.py:135]:
```suggestion
            # When there are no replicas but queries are queued, scale up the replicas
```

File: python/ray/serve/constants.py [hunk 41]
```diff
@@ -96,6 +96,9 @@
     "[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89aAbB][a-f0-9]{3}-[a-f0-9]{12}"
 )
 
+# Handle metric push interval. (the interval will indicate the cold start time period)
+HANDLE_METRIC_PUSH_INTERVAL_S = 10
```
Comment (shrekris-anyscale) [inline python/ray/serve/constants.py:100]: Nice!

File: python/ray/serve/constants.py [hunk 42]
```diff
@@ -96,6 +96,9 @@
     "[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89aAbB][a-f0-9]{3}-[a-f0-9]{12}"
 )
 
+# Handle metric push interval. (the interval will indicate the cold start time period)
```
Comment (shrekris-anyscale) [inline python/ray/serve/constants.py:99]:
```suggestion
# Handle metric push interval. This interval will affect the cold start time period.
```

File: python/ray/serve/controller.py (hunk 21 above)
Comment (shrekris-anyscale) [reply python/ray/serve/controller.py:132]: That sounds good!

File: python/ray/serve/deployment_state.py (hunk 25 above)
Comment (shrekris-anyscale) [reply python/ray/serve/deployment_state.py:1338]: Ah, good point, the `if` makes sense.

File: python/ray/serve/deployment_state.py [hunk 43]
```diff
@@ -1330,12 +1331,11 @@ def _check_curr_status(self) -> bool:
             )
             == 0
         ):
-            # Check for deleting.
-            if target_replica_count == 0 and all_running_replica_cnt == 0:
+            if self._deleting:
```
Comment (shrekris-anyscale) [inline python/ray/serve/deployment_state.py:1334]: Are we checking if `all_running_replica_cnt == 0`? t looks like we're only checking that `target_replica_count == 0` (since that's where we set `self._deleting = True`), meaning that the deployment might not be completely deleted yet.

File: python/ray/serve/handle.py (hunk 28 above)
Comment (shrekris-anyscale) [reply python/ray/serve/handle.py:221]: Got it, that sounds good.

**Review by shrekris-anyscale — CHANGES_REQUESTED** [review]
Nice work on all the updates! I left a few more comments, mostly to address some formatting nits.

File: python/ray/serve/autoscaling_metrics.py (hunk 36 above)
Comment (sihanwang41) [reply python/ray/serve/autoscaling_metrics.py:120]: Following the old implementation `TimeStampedValue`. (this is not a change in this pr)

File: python/ray/serve/autoscaling_metrics.py (hunk 38 above)
Comment (sihanwang41) [reply python/ray/serve/autoscaling_metrics.py:168]: following the window_average implementation, 

points_after_idx need to be iterable (None is not iterable)

File: python/ray/serve/deployment_state.py (hunk 43 above)
Comment (sihanwang41) [reply python/ray/serve/deployment_state.py:1334]: we are not going to delete the deployment when all_running_replica_cnt == 0 only. We will delete the deployment only when delete_deployment is called.

File: python/ray/serve/autoscaling_metrics.py (hunk 39 above)
Comment (sihanwang41) [reply python/ray/serve/autoscaling_metrics.py:155]: I think we don't add space in most of docstrings in the code. :) Please let me know if we are under any style change.

File: python/ray/serve/autoscaling_metrics.py (hunk 38 above)
Comment (sihanwang41) [reply python/ray/serve/autoscaling_metrics.py:168]: points_after_idx need to be iterable. I think that is why we have 
```
if len(points_after_idx) == 0:
    return
```

File: python/ray/serve/deployment_state.py (hunk 43 above)
Comment (shrekris-anyscale) [reply python/ray/serve/deployment_state.py:1334]: I see, that makes sense since a deployment can still be alive and have 0 replicas. My concern is that this function's return value indicates whether a deployment is finished deleting, not that it's in the process of deleting. Should we check `if self._deleting and all_running_replica_cnt == 0` here? Otherwise, a deployment could have `self._deleting=True`, but still have replicas running when we return True.

File: python/ray/serve/autoscaling_metrics.py (hunk 38 above)
Comment (shrekris-anyscale) [reply python/ray/serve/autoscaling_metrics.py:169]: Yeah, `points_after_idx` should be iterable. My suggestion is to replace the `if len(points_after_idx) == 0` check with the `default` flag in max.

File: python/ray/serve/autoscaling_metrics.py (hunk 39 above)
Comment (shrekris-anyscale) [reply python/ray/serve/autoscaling_metrics.py:156]: Hmm, yeah we should make a more consistent style guide for this. I suggested it because in some places in this change, [there's a space](https://github.com/ray-project/ray/pull/24892/files#diff-d59bb9bd79ea8faac34971edc10b2ceab555f660c0c7a663d5ec9d86fa416886R83) before the type and in others (like here) there's not.

If there's a lot of places to edit for this, we don't need to spend too much time on this. Feel free to resolve this conversation.

File: python/ray/serve/autoscaling_metrics.py (hunk 36 above)
Comment (shrekris-anyscale) [reply python/ray/serve/autoscaling_metrics.py:121]: Got it, we can refactor this in a follow-up change.

File: python/ray/serve/deployment_state.py (hunk 43 above)
Comment (sihanwang41) [reply python/ray/serve/deployment_state.py]: good point, added it

File: python/ray/serve/autoscaling_metrics.py (hunk 39 above)
Comment (sihanwang41) [reply python/ray/serve/autoscaling_metrics.py:157]: I will make further changes until we have consistent style, i think code base is not clean for this consistency.

**Review by shrekris-anyscale — APPROVED** [review]
Great work on the change!

**Comment by zhe-thoughts** [comment]
Exciting feature!

Does this warrant docs changes? Should we let users know that this now becomes possible?

**Comment by simon-mo** [comment]
Yes @sihanwang41 is following up with the doc changes along side productionize Serve autoscaling feature this month. We will tell users to try it out with 2.0.0beta release.
